File: CRM/backend/routes/hcps.js
const path = require('path');
const express = require('express');
const { stringify } = require('csv-stringify/sync');
const { Op, ValidationError, UniqueConstraintError, fn, col, where: whereFn } = require('sequelize');
const Hcp = require('../models/hcp');
const { hasAnyRole } = require('../middleware/auth');
const { importHcpsFromFile, importHcps } = require('../services/hcpsImport');

const router = express.Router();

const DATA_FILE = path.join(__dirname, '..', 'data', 'CLIENTLIST-DOPAMINE.xlsx');
const MANAGER_ROLES = ['sales_manager'];
const DEFAULT_PAGE = 1;
const DEFAULT_PAGE_SIZE = 25;
// Allow reasonably large pages so the UI can fetch all accounts when needed.
// Response shape:
// {
//   data: [ ...HCP rows... ],
//   pagination: { page, pageSize, total, totalPages }
// }
const MAX_PAGE_SIZE = 5000;

const serialize = model => ({
  id: model.id,
  name: model.name,
  areaTag: model.areaTag,
  specialty: model.specialty,
  city: model.city,
  area: model.area,
  segment: model.segment,
  phone: model.phone,
  mobile: model.mobile,
  email: model.email,
  createdAt: model.createdAt,
  updatedAt: model.updatedAt,
});

const normalizePayload = payload => ({
  name: typeof payload.name === 'string' ? payload.name.trim() : payload.name,
  areaTag: typeof payload.areaTag === 'string' ? payload.areaTag.trim() || null : payload.areaTag,
  specialty: typeof payload.specialty === 'string' ? payload.specialty.trim() || null : payload.specialty,
  city: typeof payload.city === 'string' ? payload.city.trim() || null : payload.city,
  area: typeof payload.area === 'string' ? payload.area.trim() || null : payload.area,
  segment: typeof payload.segment === 'string' ? payload.segment.trim() || null : payload.segment,
  phone: typeof payload.phone === 'string' ? payload.phone.trim() || null : payload.phone,
  mobile: typeof payload.mobile === 'string' ? payload.mobile.trim() || null : payload.mobile,
  email: typeof payload.email === 'string' ? payload.email.trim().toLowerCase() || null : payload.email,
});

const handleSequelizeError = (error, res) => {
  if (error instanceof UniqueConstraintError) {
    res.status(409).json({ message: 'An HCP with the same identity already exists.' });
    return true;
  }

  if (error instanceof ValidationError) {
    res.status(400).json({ message: error.message });
    return true;
  }

  return false;
};

const parsePositiveInt = (value, fallback) => {
  const parsed = Number.parseInt(value, 10);
  if (!Number.isInteger(parsed) || parsed <= 0) {
    return fallback;
  }
  return parsed;
};

const buildFilters = query => {
  const where = {};
  const andConditions = [];

  const likeValue = value => ({ [Op.like]: `%${value}%` });

  if (typeof query.city === 'string' && query.city.trim()) {
    where.city = likeValue(query.city.trim());
  }

  if (typeof query.area === 'string' && query.area.trim()) {
    where.area = likeValue(query.area.trim());
  }

  if (typeof query.specialty === 'string' && query.specialty.trim()) {
    where.specialty = likeValue(query.specialty.trim());
  }

  if (typeof query.segment === 'string' && query.segment.trim()) {
    where.segment = likeValue(query.segment.trim());
  }

  if (typeof query.search === 'string') {
    const normalized = query.search.trim().toLowerCase();
    if (normalized) {
      andConditions.push({
        [Op.or]: [
          whereFn(fn('lower', col('name')), { [Op.like]: `%${normalized}%` }),
          whereFn(fn('lower', col('city')), { [Op.like]: `%${normalized}%` }),
          whereFn(fn('lower', col('area')), { [Op.like]: `%${normalized}%` }),
          whereFn(fn('lower', col('specialty')), { [Op.like]: `%${normalized}%` }),
        ],
      });
    }
  }

  if (andConditions.length) {
    where[Op.and] = andConditions;
  }

  return where;
};

const ensureManagerAccess = (req, res) => {
  if (!hasAnyRole(req.user, MANAGER_ROLES)) {
    res.status(403).json({ message: 'Insufficient permissions.' });
    return false;
  }
  return true;
};

router.get('/', async (req, res, next) => {
  const page = parsePositiveInt(req.query.page, DEFAULT_PAGE);
  const requestedPageSize = parsePositiveInt(req.query.pageSize, DEFAULT_PAGE_SIZE);
  const pageSize = Math.min(requestedPageSize, MAX_PAGE_SIZE);
  const where = buildFilters(req.query || {});

  try {
    const { rows, count } = await Hcp.findAndCountAll({
      where,
      limit: pageSize,
      offset: (page - 1) * pageSize,
      order: [
        ['name', 'ASC'],
        ['id', 'ASC'],
      ],
    });

    const totalPages = Math.max(1, Math.ceil(count / pageSize));
    res.json({
      data: rows.map(serialize),
      pagination: {
        page,
        pageSize,
        total: count,
        totalPages,
      },
    });
  } catch (error) {
    next(error);
  }
});

router.get('/export', async (req, res, next) => {
  if (!ensureManagerAccess(req, res)) {
    return;
  }

  try {
    const hcps = await Hcp.findAll({
      order: [
        ['city', 'ASC'],
        ['area', 'ASC'],
        ['name', 'ASC'],
      ],
    });

    const records = hcps.map(serialize);
    const csv = stringify(records, {
      header: true,
      columns: [
        { key: 'name', header: 'Name' },
        { key: 'specialty', header: 'Specialty' },
        { key: 'city', header: 'City' },
        { key: 'area', header: 'Area' },
        { key: 'segment', header: 'Segment' },
        { key: 'phone', header: 'Phone' },
        { key: 'mobile', header: 'Mobile' },
        { key: 'email', header: 'Email' },
        { key: 'areaTag', header: 'Area Tag' },
      ],
    });

    res.setHeader('Content-Type', 'text/csv');
    res.setHeader('Content-Disposition', 'attachment; filename="hcps.csv"');
    res.send(csv);
  } catch (error) {
    next(error);
  }
});

router.post('/import', async (req, res, next) => {
  if (!ensureManagerAccess(req, res)) {
    return;
  }

  try {
    const rows = Array.isArray(req.body?.rows) ? req.body.rows : null;
    const summary = rows ? await importHcps(rows) : await importHcpsFromFile(DATA_FILE);
    res.json(summary);
  } catch (error) {
    next(error);
  }
});

router.get('/:id', async (req, res, next) => {
  try {
    const hcp = await Hcp.findByPk(req.params.id);
    if (!hcp) {
      return res.status(404).json({ message: 'HCP not found.' });
    }

    res.json(serialize(hcp));
  } catch (error) {
    next(error);
  }
});

router.post('/', async (req, res, next) => {
  try {
    const payload = normalizePayload(req.body || {});
    const hcp = await Hcp.create(payload);
    res.status(201).json(serialize(hcp));
  } catch (error) {
    if (!handleSequelizeError(error, res)) {
      next(error);
    }
  }
});

router.put('/:id', async (req, res, next) => {
  try {
    const hcp = await Hcp.findByPk(req.params.id);
    if (!hcp) {
      return res.status(404).json({ message: 'HCP not found.' });
    }

    const payload = normalizePayload(req.body || {});
    await hcp.update(payload);
    res.json(serialize(hcp));
  } catch (error) {
    if (!handleSequelizeError(error, res)) {
      next(error);
    }
  }
});

router.delete('/:id', async (req, res, next) => {
  try {
    const hcp = await Hcp.findByPk(req.params.id);
    if (!hcp) {
      return res.status(404).json({ message: 'HCP not found.' });
    }

    await hcp.destroy();
    res.status(204).send();
  } catch (error) {
    next(error);
  }
});

module.exports = router;

File: ALQASEER-PWA/CRM/backend/api/v1/hcps.py
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import func, or_
from sqlalchemy.orm import Session

from core.db import get_db
from core.security import get_current_user, require_roles
from models.hcp import HCP

router = APIRouter(
    prefix="/hcps",
    tags=["hcps"],
    dependencies=[Depends(get_current_user)],
)

DEFAULT_PAGE = 1
DEFAULT_PAGE_SIZE = 25
MAX_PAGE_SIZE = 5000


def _serialize_hcp(hcp: HCP) -> dict:
    name = f"{hcp.first_name} {hcp.last_name}".strip()
    return {
        "id": hcp.id,
        "name": name,
        "areaTag": hcp.area,
        "specialty": hcp.specialty,
        "city": hcp.city,
        "area": hcp.area,
        "segment": None,
        "phone": hcp.phone,
        "mobile": None,
        "email": hcp.email,
        "createdAt": hcp.created_at,
        "updatedAt": hcp.updated_at,
    }


def _apply_filters(
    query,
    *,
    search: Optional[str],
    area_tag: Optional[str],
    specialty: Optional[str],
    segment: Optional[str],
):
    if search:
        term = f"%{search.strip().lower()}%"
        query = query.filter(
            or_(
                func.lower(HCP.first_name).like(term),
                func.lower(HCP.last_name).like(term),
                func.lower(HCP.specialty).like(term),
                func.lower(HCP.area).like(term),
                func.lower(HCP.city).like(term),
            )
        )
    if area_tag:
        term = f"%{area_tag.strip().lower()}%"
        query = query.filter(func.lower(HCP.area).like(term))
    if specialty:
        term = f"%{specialty.strip().lower()}%"
        query = query.filter(func.lower(HCP.specialty).like(term))
    if segment:
        term = f"%{segment.strip().lower()}%"
        query = query.filter(func.lower(HCP.area).like(term))
    return query


@router.get("", response_model=dict)
def list_hcps(
    page: int = Query(DEFAULT_PAGE, ge=1),
    page_size: int = Query(DEFAULT_PAGE_SIZE, ge=1, le=MAX_PAGE_SIZE, alias="pageSize"),
    search: Optional[str] = None,
    area_tag: Optional[str] = Query(default=None, alias="areaTag"),
    specialty: Optional[str] = None,
    segment: Optional[str] = None,
    db: Session = Depends(get_db),
):
    query = db.query(HCP).filter(HCP.is_active.is_(True))
    query = _apply_filters(
        query,
        search=search,
        area_tag=area_tag,
        specialty=specialty,
        segment=segment,
    )

    total = query.count()
    items = (
        query.order_by(HCP.last_name.asc(), HCP.first_name.asc())
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )

    total_pages = (total + page_size - 1) // page_size if total else 0
    data = [_serialize_hcp(item) for item in items]
    return {
        "data": data,
        "meta": {
            "page": page,
            "pageSize": page_size,
            "total": total,
            "totalPages": total_pages,
        },
    }


@router.get("/{hcp_id}", response_model=dict)
def get_hcp(hcp_id: int, db: Session = Depends(get_db)) -> dict:
    hcp = db.query(HCP).filter(HCP.id == hcp_id, HCP.is_active.is_(True)).first()
    if not hcp:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="HCP not found.")
    return _serialize_hcp(hcp)


@router.post(
    "",
    response_model=dict,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_roles("sales_manager", "admin"))],
)
def create_hcp(
    payload: dict,
    db: Session = Depends(get_db),
):
    name = str(payload.get("name") or "").strip()
    if not name:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Name is required.")
    parts = name.split()
    first_name = parts[0]
    last_name = " ".join(parts[1:]) if len(parts) > 1 else "-"

    hcp = HCP(
        first_name=first_name,
        last_name=last_name,
        specialty=payload.get("specialty"),
        phone=payload.get("phone"),
        email=payload.get("email"),
        clinic_address=payload.get("clinic_address"),
        area=payload.get("area"),
        city=payload.get("city"),
    )
    db.add(hcp)
    db.commit()
    db.refresh(hcp)
    return _serialize_hcp(hcp)


@router.put(
    "/{hcp_id}",
    response_model=dict,
    dependencies=[Depends(require_roles("sales_manager", "admin"))],
)
def update_hcp(
    hcp_id: int,
    payload: dict,
    db: Session = Depends(get_db),
):
    hcp = db.query(HCP).filter(HCP.id == hcp_id).first()
    if not hcp:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="HCP not found.")

    if "name" in payload and payload.get("name"):
        parts = str(payload["name"]).strip().split()
        hcp.first_name = parts[0]
        hcp.last_name = " ".join(parts[1:]) if len(parts) > 1 else "-"
    for key in ("specialty", "phone", "email", "clinic_address", "area", "city"):
        if key in payload:
            setattr(hcp, key, payload.get(key))

    db.commit()
    db.refresh(hcp)
    return _serialize_hcp(hcp)


@router.delete(
    "/{hcp_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_roles("sales_manager", "admin"))],
)
def delete_hcp(hcp_id: int, db: Session = Depends(get_db)):
    hcp = db.query(HCP).filter(HCP.id == hcp_id).first()
    if not hcp or not hcp.is_active:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="HCP not found.")
    hcp.is_active = False
    db.commit()
